### DISCUSSION TITLE
Batch Size on stdout.write

### DISCUSSION THREAD

**Discussion by DylanBPY** [pybricks/support#1870] [Q&A] [answered]
Hi, it's me again! Sorry for all the posts, trying to get a really big program to work, I'll share that here when it's done. Anyways, I'm trying to work on sending data from PC to Spike, I used the docs as a reference and now I have this handle_rx function:
```
    async def handle_rx(_, rx_data: bytearray):
        global setup, initial_data
        try:
            if rx_data[0] == 0x01: # "write stdout" event (0x01)
                rx_data = rx_data[1:] # Remove the event type
                if rx_data.startswith(b"#DATA#"):
                    # Strip the #DATA# prefix directly from the encoded bytes
                    payload_encoded = rx_data[len("#DATA#"):]  # Remove #DATA# prefix from the bytearra

                    if initial_data is None:
                        payload_decoded = payload_encoded.decode()  # Decode the remaining bytes
                        print(f"Received initial data chunk: {payload_decoded}")
                        done = add_chunk_to_data(payload_encoded)

                        if not setup and done:
                            print("Received all initial data, setting up visualization...")
                            #SetupVisualization()
                            print("Visualization set up.")
                            setup = True
                            '''
                            await client.write_gatt_char(
                                PYBRICKS_COMMAND_EVENT_CHAR_UUID,
                                b"\x06ok",  # prepend "write stdin" command (0x06) and "ok" response
                                response=True
                            )
                            '''
                            print("Path received and acknowledged, starting visualization...")
                        else:
                            pass
                    else:
                        decoded_data = struct.unpack('6f', payload_encoded)
                        print(f"Received data chunk: {decoded_data}")
                        #UpdateVisualization(*decoded_data)
                else:
                    print(f"Received regular print output: {rx_data.decode()}")
        except Exception as e:
            print(f"Error in handle_rx: {e}")
            terminate_thread = True
```
(Sorry if this code is a little bit messed up, I'm on mobile right now)

I've got the actual data sending part working, but the problem is, if for example I wanted to send this:
`Path sent to visualization server. Waiting for acknowledgement...`
It would turn into this:
```
Received regular print output: Path sent t
Received regular print output: o visualization ser
Received regular print output: ver. Waiting for se
Received regular print output: rver acknowledgment
Received regular print output: ...
```
It looks like there's a buffer of some sort, is there a way I could make the buffer a set amount or remove it completely?

**Comment by dlech** [comment, marked as answer]
No, there isn't a way to change this (at least without modifying the firmware).

If you want to send variable sized messages, I would suggest to devise a protocol that includes the size of the message followed by the message so that the receiving end can know where to break things up into separate messages.

Or you could devise a protocol with fixed-sized messages of 20 bytes or less and have the receiving end translate that into something more human readable. For example send "STATE:PATH SENT" and have a lookup table on the PC to translate that to your longer message.

**Reply by DylanBPY** [reply]
Is data separated into 20 byte chunks?

**Reply by DylanBPY** [reply]
In other words, no matter how much data is sent via stdout.write, it will separate said data in chunks of 20 bytes on the server side?

**Reply by dlech** [reply]
Yes, that is the current implementation, at least when you write more than 20 bytes. If you are writing less that 20 bytes and there is sufficient delay between writes, the writes won't be combined.

**Reply by DylanBPY** [reply]
Alright, ill try and fix the program now, thanks!
